### PR TITLE
Froidure pin hpp

### PR DESCRIPTION
**Pull request by james-d-mitchell**

### CODE REVIEW

File: tests/test-froidure-pin-intmat.cpp
Comment (james-d-mitchell) [reply tests/test-froidure-pin-intmat.cpp:1]: The tests now pass for me, is this still an issue for you @Joseph-Edwards ?

Comment (Joseph-Edwards) [reply tests/test-froidure-pin-intmat.cpp:1]: To be explicit, this new test
```diff
 LIBSEMIGROUPS_TEST_CASE("FroidurePin<IntMat<0, int64_t>>",
                         "033",
                         "Example 000",
                         "[quick][froidure-pin][intmat]") {
   test_IntMat000<IntMat<0, 0, int64_t>>();
+  test_IntMat000<IntMat<>>();
  }
```
gives this output
```rst
Randomness seeded to: 1513999630
[033]: FroidurePin<IntMat<0, int64_t>> 033: Example 000            
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_froidure_pin is a Catch2 v3.7.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
FroidurePin<IntMat<0, int64_t>> 033: Example 000
-------------------------------------------------------------------------------
tests/test-froidure-pin-intmat.cpp:218
...............................................................................

tests/test-froidure-pin-intmat.cpp:61: FAILED:
  REQUIRE( S.current_size() == 631 )
with expansion:
  311 (0x137) == 631 (0x277)

............
===============================================================================
Total time ....................................................................
===============================================================================
test cases:   1 |   0 passed | 1 failed
assertions: 647 | 646 passed | 1 failed
```

Comment (james-d-mitchell) [reply tests/test-froidure-pin-intmat.cpp:1]: Ah right yes I'd expect that to fail, seems ok. I wanna look over the changes in this pr before merging, so please don't merge just yet

Comment (Joseph-Edwards) [reply tests/test-froidure-pin-intmat.cpp:1]: Likewise. I still need to look over the doc too. I'll do a review at some point next week

**Comment by james-d-mitchell** [comment]
This is now complete AFAI concerned. Still need to update the python bindings

**Comment by Joseph-Edwards** [comment]
@james-d-mitchell do you have any uncommitted changes for this, or shall I start reviewing? I'm happy to fix these tests that are trivially failing for `to_human_readable_repr` reasons but don't want to fix anything you've already done.

**Comment by james-d-mitchell** [comment]
@Joseph-Edwards I'm done with this, happy if you can fix the tests, also happy to do it myself

**Review by Joseph-Edwards — COMMENTED** [review]
Modulo the minor changes outlined below, this looks good to me. Thanks @james-d-mitchell!

File: include/libsemigroups/bipart.hpp
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:768]: The parameter `braces` should be documented.

File: include/libsemigroups/bipart.hpp
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:1484]: The parameter `braces` should be documented.

**Review by Joseph-Edwards — APPROVED** [review]
All looks great now. Thanks @james-d-mitchell!

I'll hold off on merging whilst you're still working on the Python bindings in case you realise that there's anything else that needs a change.

**Pull request merged**
